rust/src/builtins/apply_live.rs

```diff
@@ -0,0 +1,106 @@
+//!
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+use crate::cxxrsutil::*;
+use crate::live;
+use anyhow::{anyhow, Result};
+use gio::DBusProxyExt;
+use glib::translate::*;
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+#[structopt(name = "apply-live")]
+#[structopt(rename_all = "kebab-case")]
+struct Opts {
+    /// Target provided commit instead of pending deployment
+    target: Option<String>,
+    /// Reset back to booted commit
+    #[structopt(long)]
+    reset: bool,
+}
+
+fn require_booted_deployment(sysroot: &ostree::Sysroot) -> Result<ostree::Deployment> {
+    sysroot
+        .get_booted_deployment()
+        .ok_or_else(|| anyhow!("Not booted into an OSTree system"))
+}
+
+fn get_args_variant(opts: &Opts) -> Result<glib::Variant> {
+    let r = glib::VariantDict::new(None);
+
+    if let Some(target) = opts.target.as_ref() {
+        if opts.reset {
+            return Err(anyhow!("Cannot specify both --target and --reset"));
+        }
+        r.insert("target", &target.as_str());
+    } else if opts.reset {
+        let sysroot = &ostree::Sysroot::new_default();
+        sysroot.load(gio::NONE_CANCELLABLE)?;
+        let booted = require_booted_deployment(sysroot)?;
+        // Unwrap safety: This can't return NULL
+        let csum = booted.get_csum().expect("csum");
+        r.insert("target", &csum.as_str());
+    }
+
+    Ok(r.end())
+}
+
+pub(crate) fn applylive_entrypoint(args: &Vec<String>) -> Result<()> {
+    let opts = Opts::from_iter(args.iter());
+    let client = &mut crate::client::ClientConnection::new()?;
+    let args = get_args_variant(&opts)?;
+
+    // FIXME glib variant bindings
+    let params = unsafe {
+        let args = args.to_glib_none().0;
+        let r = glib_sys::g_variant_new_tuple(&args as *const *mut _, 1);
+        glib_sys::g_variant_ref_sink(r);
+        from_glib_full(r)
+    };
```

Comment:
Might be worth a wrapper function for this bit.

Reply:
OK yeah, definitely.  I added a `variant_utils.rs`.